pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,75 +87,127 @@ public static ControllerLeaderLocator getInstance() {
+            String controllerInstanceId = LeadControllerUtils.extractControllerInstanceId(participantInstanceId);
+            Pair<String, Integer> leadControllerPair = convertToHostAndPortPair(controllerInstanceId);
+            masterFound = true;
+            _cachedControllerLeaderMap.put(partitionId, leadControllerPair);
+          }
+        }
+        if (!masterFound) {
+          // It's ok to log a warning since we can be in this state for some small time during the migration.
+          // Otherwise, we are attempted to mark this as an error.
+          LOGGER.warn("There is no controller in MASTER state for partition: {} in {}", partitionName,
+              Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+          return;
+        }
+      }
+      LOGGER.info("Refreshed controller leader map successfully.");
+      refreshSucceeded = true;
     } catch (Exception e) {
-      LOGGER.warn("Could not get whether lead controller resource is enabled or not.", e);
-      return false;
+      LOGGER.warn("Caught exception when getting lead controller instance Id from external view of {}",
+          Helix.LEAD_CONTROLLER_RESOURCE_NAME, e);
+    } finally {
+      _cachedControllerLeaderValid = refreshSucceeded;
     }
   }
 
   /**
-   * Gets leader from lead controller resource. Null if there is no leader.
-   * @param rawTableName raw table name.
-   * @return pair of instance hostname and port of Helix cluster leader, e.g. {localhost, 9000}.
+   * Updates lead controller pairs from Helix cluster leader.
    */
-  private Pair<String, Integer> getLeaderFromLeadControllerResource(String rawTableName) {
-    Pair<String, Integer> leaderHostAndPortPair = getLeadControllerInstanceIdForTable(rawTableName);
-    if (leaderHostAndPortPair != null) {
-      return leaderHostAndPortPair;
-    } else {
-      LOGGER.warn("Could not locate leader for table: {}", rawTableName);
-      return null;
+  private void refreshControllerLeaderMapFromHelixClusterLeader() {
+    boolean refreshSucceeded = false;
+    try {
+      Pair<String, Integer> helixClusterLeader = getHelixClusterLeader();
+      if (helixClusterLeader == null) {
+        LOGGER.error("Failed to refresh the controller leader map.");
+        return;
+      }
+      for (int i = 0; i < Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE; i++) {
+        _cachedControllerLeaderMap.put(i, helixClusterLeader);
+      }
+      refreshSucceeded = true;
+      LOGGER.info("Refreshed controller leader map successfully.");
+    } finally {
+      _cachedControllerLeaderValid = refreshSucceeded;
+    }
+  }
+
+  /**
+   * Checks whether lead controller resource is enabled or not. The switch is in resource config.
+   */
+  private boolean isLeadControllerResourceEnabled() {
+    BaseDataAccessor<ZNRecord> dataAccessor = _helixManager.getHelixDataAccessor().getBaseDataAccessor();
+    Stat stat = new Stat();
+    try {
+      ZNRecord znRecord = dataAccessor
```

Comment:
We don't have a Helix util method for this? (Same for controller leader node)

Reply:
There are two ways to get the helix leader. One is presented in this PR. Another is to fetch live instance from `helixDataAccessor`.
Both controller and server need to fetch helix leader. I have another PR which puts this logic from locator to utils. At that time, we can merge this together to unify the logic.